contracts/filters/MerkleCollateralFilter.sol

```diff
@@ -0,0 +1,188 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity 0.8.20;
+
+import "@openzeppelin/contracts/utils/cryptography/MerkleProof.sol";
+
+import "../CollateralFilter.sol";
+
+/**
+ * @title Merkle Collateral Filter
+ * @author MetaStreet Labs
+ */
+contract MerkleCollateralFilter is CollateralFilter {
+    /**************************************************************************/
+    /* Errors */
+    /**************************************************************************/
+
+    /**
+     * @notice Invalid root
+     */
+    error InvalidRoot();
+
+    /**
+     * @notice Invalid node count
+     */
+    error InvalidNodeCount();
+
+    /**
+     * @notice Invalid context
+     */
+    error InvalidContext();
+
+    /**************************************************************************/
+    /* State */
+    /**************************************************************************/
+
+    /**
+     * @notice Supported token
+     */
+    address private _token;
+
+    /**
+     * @notice Number of nodes for each proof
+     */
+    uint32 private _nodeCount;
+
+    /**
+     * @notice Merkle root
+     */
+    bytes32 private _root;
+
+    /**
+     * @notice Metadata URI
+     */
+    string private _metadataURI;
+
+    /**************************************************************************/
+    /* Initializer */
+    /**************************************************************************/
+
+    /**
+     * @notice MerkleCollateralFilter initializer
+     */
+    function _initialize(address token, bytes memory params) internal {
+        /* Decode parameters */
+        (bytes32 root, uint32 nodeCount, string memory metadataURI_) = abi.decode(params, (bytes32, uint32, string));
```

Comment:
We should just take these parameters in directly here and decode once in Pool configuration `initialize()` (right now it's being decoded, re-encoded, and re-decoded)